curve-amm/src/lib.rs

```diff
@@ -0,0 +1,1062 @@
+        let mut balances = Vec::new();
+        for asset in assets {
+            let reply: Event = msg::send_and_wait_for_reply(
+                *asset,
+                &Action::BalanceOf(exec::program_id()),
+                100_000_000_000,
+                0,
+            )
+            .await
+            .expect("Error in async message");
+            let asset_balance = match reply {
+                Event::Balance(bal) => FixedU128::saturating_from_integer(bal),
+                _ => {
+                    panic!("could not decode TotalIssuance reply");
+                }
+            };
+            balances.push(asset_balance);
+        }
+        balances
+    }
+
+    async fn transfer_funds_to_pool(who: &ActorId, amounts: Vec<FixedU128>, assets: &[ActorId]) {
+        let zero = FixedU128::zero();
+        for (i, amount) in amounts.iter().enumerate() {
+            if amount > &zero {
+                let amount_u = amount.into_inner() / FixedU128::DIV;
+                let _reply: Event = msg::send_and_wait_for_reply(
+                    assets[i],
+                    &Action::TransferFrom(TransferFromInput {
+                        owner: *who,
+                        to: exec::program_id(),
+                        amount: amount_u,
+                    }),
+                    100_000_000_000,
+                    0,
+                )
+                .await
+                .expect("could not decode Transfer reply");
+            }
+        }
+    }
+
+    async fn transfer_funds_from_pool(who: &ActorId, amounts: Vec<FixedU128>, assets: &[ActorId]) {
+        let zero = FixedU128::zero();
+        for (i, amount) in amounts.iter().enumerate() {
+            if amount > &zero {
+                let amount: u128 = Self::fixed_to_u128(amount);
+                let _reply: Event = msg::send_and_wait_for_reply(
+                    assets[i],
+                    &Action::Transfer(TransferInput { to: *who, amount }),
+                    100_000_000_000,
+                    0,
+                )
+                .await
+                .expect("could not decode Transfer reply");
+            }
+        }
+    }
+
+    pub fn get_pool(&self, pool_id: &PoolId) -> &PoolInfo {
+        self.pools.get(pool_id).expect("get_pool: Pool Not Found")
+    }
+
+    pub async fn get_lp_token_suppy(&self, pool_id: &PoolId) -> FixedU128 {
+        let pool = self.get_pool(pool_id);
+        let reply: Event =
+            msg::send_and_wait_for_reply(pool.pool_asset, &Action::TotalSupply, 100_000_000_000, 0)
+                .await
+                .expect("Error in async message");
+        let token_supply = match reply {
+            Event::TotalSupply(bal) => FixedU128::saturating_from_integer(bal),
+            _ => {
+                panic!("could not decode TotalIssuance reply");
+            }
+        };
+        token_supply
+    }
+
+    #[allow(dead_code)]
+    async fn add_liquidity(
```

Comment:
I'd rewrite this function (and similar) by returning `Result`. This allows avoiding a lot of `mem::drop` as well as `match option` (one may use `option.ok_or()` function instead). You can call `panic!` outside this function.